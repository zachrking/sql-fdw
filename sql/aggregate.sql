--SET log_min_messages  TO DEBUG1;
--SET client_min_messages  TO DEBUG1;
CREATE EXTENSION sqlite_fdw;
CREATE SERVER sqlite_svr FOREIGN DATA WRAPPER sqlite_fdw
OPTIONS (database '/tmp/sqlitefdw_test.db');
CREATE FOREIGN TABLE multiprimary(a int, b int OPTIONS (key 'true'), c int OPTIONS(key 'true')) SERVER sqlite_svr;
-- test for aggregate pushdown
explain (costs off, verbose) select count(distinct a) from multiprimary;

explain (costs off, verbose) select sum(b),max(b), min(b), avg(b) from multiprimary;

explain (costs off, verbose) select sum(b+5)+2 from multiprimary group by b/2 order by b/2;

explain (costs off, verbose) select sum(a) from multiprimary group by b having sum(a) > 0;

explain (costs off, verbose) select sum(a) from multiprimary group by b having avg(a^2) > 0 and sum(a) > 0;

-- stddev and variance are not pushed down
explain (costs off, verbose) select stddev(a) from multiprimary;
explain (costs off, verbose) select sum(a) from multiprimary group by b having variance(a) > 0;

DROP FOREIGN TABLE multiprimary;
DROP SERVER sqlite_svr;
DROP EXTENSION sqlite_fdw CASCADE;
