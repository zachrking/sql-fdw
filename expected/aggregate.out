--SET log_min_messages  TO DEBUG1;
--SET client_min_messages  TO DEBUG1;
CREATE EXTENSION sqlite_fdw;
CREATE SERVER sqlite_svr FOREIGN DATA WRAPPER sqlite_fdw
OPTIONS (database '/tmp/sqlitefdw_test.db');
CREATE FOREIGN TABLE multiprimary(a int, b int OPTIONS (key 'true'), c int OPTIONS(key 'true')) SERVER sqlite_svr;
-- test for aggregate pushdown
explain (costs off, verbose) select count(distinct a) from multiprimary;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Foreign Scan
   Output: (count(DISTINCT a))
   SQLite query: SELECT count(DISTINCT `a`) FROM main."multiprimary"
(3 rows)

explain (costs off, verbose) select sum(b),max(b), min(b), avg(b) from multiprimary;
                                       QUERY PLAN                                       
----------------------------------------------------------------------------------------
 Foreign Scan
   Output: (sum(b)), (max(b)), (min(b)), (avg(b))
   SQLite query: SELECT sum(`b`), max(`b`), min(`b`), avg(`b`) FROM main."multiprimary"
(3 rows)

explain (costs off, verbose) select sum(b+5)+2 from multiprimary group by b/2 order by b/2;
                                                          QUERY PLAN                                                          
------------------------------------------------------------------------------------------------------------------------------
 Foreign Scan
   Output: ((sum((b + 5)) + 2)), ((b / 2))
   SQLite query: SELECT (sum((`b` + 5)) + 2), (`b` / 2) FROM main."multiprimary" GROUP BY 2 ORDER BY (`b` / 2) ASC NULLS LAST
(3 rows)

explain (costs off, verbose) select sum(a) from multiprimary group by b having sum(a) > 0;
                                            QUERY PLAN                                            
--------------------------------------------------------------------------------------------------
 Foreign Scan
   Output: (sum(a)), b
   SQLite query: SELECT sum(`a`), `b` FROM main."multiprimary" GROUP BY 2 HAVING ((sum(`a`) > 0))
(3 rows)

explain (costs off, verbose) select sum(a) from multiprimary group by b having avg(a^2) > 0 and sum(a) > 0;
                                                              QUERY PLAN                                                               
---------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate
   Output: sum(a), b
   Group Key: multiprimary.b
   Filter: ((avg(((multiprimary.a)::double precision ^ '2'::double precision)) > '0'::double precision) AND (sum(multiprimary.a) > 0))
   ->  Foreign Scan on public.multiprimary
         Output: a, b, c
         SQLite query: SELECT `a`, `b` FROM main."multiprimary" ORDER BY `b` ASC NULLS LAST
(7 rows)

-- stddev and variance are not pushed down
explain (costs off, verbose) select stddev(a) from multiprimary;
                        QUERY PLAN                         
-----------------------------------------------------------
 Aggregate
   Output: stddev(a)
   ->  Foreign Scan on public.multiprimary
         Output: a, b, c
         SQLite query: SELECT `a` FROM main."multiprimary"
(5 rows)

explain (costs off, verbose) select sum(a) from multiprimary group by b having variance(a) > 0;
                                         QUERY PLAN                                         
--------------------------------------------------------------------------------------------
 GroupAggregate
   Output: sum(a), b
   Group Key: multiprimary.b
   Filter: (variance(multiprimary.a) > '0'::numeric)
   ->  Foreign Scan on public.multiprimary
         Output: a, b, c
         SQLite query: SELECT `a`, `b` FROM main."multiprimary" ORDER BY `b` ASC NULLS LAST
(7 rows)

DROP FOREIGN TABLE multiprimary;
DROP SERVER sqlite_svr;
DROP EXTENSION sqlite_fdw CASCADE;
